crates/revm/revm-inspectors/src/tracing/builder/geth.rs

```diff
@@ -238,4 +280,71 @@ impl GethTraceBuilder {
             node.geth_update_account_storage(account_states, post_value);
         }
     }
+
+    fn update_storage_from_trace_diff_mode(
+        &self,
+        account_states: &mut BTreeMap<Address, AccountState>,
+        post_value: bool,
+    ) {
+        for node in self.nodes.iter() {
+            node.geth_update_account_storage_diff_mode(account_states, post_value);
+        }
+    }
+
+    fn diff_traces(
+        &self,
+        pre: &BTreeMap<Address, AccountState>,
+        post: &BTreeMap<Address, AccountState>,
+    ) -> DiffMode {
+        let mut out_diff = DiffMode::default();
+
+        for (addr, pre_state) in pre.iter() {
+            let post_state = post.get(addr).cloned().unwrap_or_default();
+
+            // Don't put created accounts or accounts that are identical to the post
+            // state into the diff.
+            if pre_state.change_type != ChangeType::Create && pre_state != &post_state {
+                let mut pre_clone = pre_state.clone();
+                pre_clone.storage.get_or_insert_with(BTreeMap::new).retain(|_, v| !v.is_zero());
+                pre_clone.storage = (pre_clone.storage.as_ref().unwrap().len() > 0)
+                    .then_some(pre_clone.storage)
+                    .unwrap_or(None);
+                out_diff.pre.insert(*addr, pre_clone);
+            }
+        }
+
+        for (addr, post_state) in post.iter() {
+            let pre_state = pre.get(addr).cloned().unwrap_or_default();
+
+            // Don't put destroyed accounts or accounts that are identical to the pre-state
+            // into the diff.
+            if post_state.change_type == ChangeType::Destroy || &pre_state == post_state {
+                continue
+            }
+
+            // The post state should only contain the fields that have changed.
+            // To do this, we clone the post state and remove the fields that
+            // are the same as the pre state.
+            let mut post_clone = post_state.clone();
+            if pre_state.balance == post_state.balance {
+                post_clone.balance = None;
+            }
+
+            if pre_state.nonce == post_state.nonce {
+                post_clone.nonce = None;
+            }
+
+            if pre_state.code == post_state.code {
+                post_clone.code = None;
+            }
+
+            post_clone.storage.get_or_insert_with(BTreeMap::new).retain(|_, v| !v.is_zero());
+            post_clone.storage = (post_clone.storage.as_ref().unwrap().len() > 0)
```

Comment:
don't like the unwrap, why is this safe?

this could be solved with an and_then as well?